core/domain/state_domain.py

```diff
@@ -2227,6 +2382,9 @@ def __init__(self, rule_type, inputs):
                 enclosed in {{...}} braces.
         """
         self.rule_type = rule_type
+        # Here, we are narrowing down the type from Mapping to Dict. Because
+        # Mapping is used just to accept the different types of allowed Dicts.
+        assert isinstance(inputs, dict)
```

Comment:
Mapping is used just to accept the different dictionaries, because `Dict[str, int]` can be a subtype of `Mapping[str, Union[str, int]` but it cannot be a subtype of `Dict[str, Union[str, int]]`. So, after accepting the `inputs` we are again narrowing down the type to `Dict`, to match the type with implementation.